locale.c

```diff
@@ -6283,21 +6283,100 @@ S_my_langinfo_i(pTHX_
             break;
         }
 
-        /* Here, it isn't a UTF-8 locale. */
+        /* Here, it isn't a UTF-8 locale.  After the #else clause is code to
+         * find the codeset (if any) from the locale name */
 
-#        else   /* mbtowc() is not available.  The chances of this code getting
-                   compiled are very small, as it is a C99 required function,
-                   and we are now requiring C99; perhaps if it is a defective
-                   implementation.  But if so, there are other libc functions
-                   that could be used instead. */
+#        else
+
+        /* Here, neither mbtowc() nor mbrtowc() is available.  The chances of
+         * this are very small, as they are C99 required functions, and we are
+         * now requiring C99; perhaps this is a defective implementation and
+         * therefore Configure has been set to indicate neither exists.
+         *
+         * Just below we try to calculate the code set from the locale name.
+         * In all cases but this one, it has already been determined that it is
+         * not a UTF-8 locale.  But for this case, we defer that, calculate the
+         * code set name, if any, and later use that result as a hint.  First
+         * #define a symbol to later tell us that we need to handle this case.
+         * */
+#          define NEED_FURTHER_UTF8NESS_CHECKING
+#        endif
+
+        /* Here, the code set has not been found.  The only other option khw
+         * could think of is to see if the codeset is part of the locale name.
+         * This is very less than ideal; often there is no code set in the
+         * name; and at other times they even lie.
+         *
+         * But there is an XPG standard syntax, which many locales follow:
+         *
+         * language[_territory[.codeset]][@modifier]
+         *
+         * So we take the part between the dot and any '@' */
+        retval = (const char *) strchr(locale, '.');
```

Comment:
Why cast here?

In C strchr() returns `char *` which trivially converts to `const char *`.

In C++, since `locale` is `const char *` so the `const char *` [overload from](https://en.cppreference.com/w/cpp/string/byte/strchr)\*:
```
const char* strchr( const char* str, int ch ); 
char* strchr( char* str, int ch );
```
is selected here, which already returns `const char *`.

There's no cast for the similar assignment to `modifier` below.

\* `#include <string.h>` in C++ behaves like including `<cstring.h>` but adds the names that would normally be added to the `std::` namespace to the [global namespace](https://en.cppreference.com/w/cpp/header#C_compatibility_headers)

Reply:
It was old code that I simply moved from elsewhere in the file without changing it.  I imagine there were different types at one time that required the case.  But now changed.